File: esoteric_entropy/sources/sysctl.py
"""Sysctl kernel counter entropy source — the crown jewel.

macOS exposes 1600+ sysctl keys.  Many are kernel counters that change
rapidly (TCP stats, VM page faults, context switches, etc.).  By sampling
the *deltas* of fluctuating keys at high frequency, we harvest entropy
from the unpredictable micro-behaviour of the entire operating system.

Discovery on a Mac Mini M4 found **58 keys** that change within 0.2 s.
"""

from __future__ import annotations

import platform
import subprocess
import time

import numpy as np

from esoteric_entropy.sources.base import EntropySource

# Categories for discovered keys
CATEGORIES = {
    "vm": "Virtual memory counters",
    "net": "Network stack counters",
    "kern": "Kernel statistics",
    "debug": "Debug / profiling counters",
    "security": "Security subsystem counters",
    "hw": "Hardware counters",
}


def _sysctl_snapshot() -> dict[str, int]:
    """Batch read ALL sysctl keys as integers using a single `sysctl -a` call."""
    try:
        r = subprocess.run(
            ["/usr/sbin/sysctl", "-a"],
            capture_output=True,
            text=True,
            timeout=10,
        )
    except (subprocess.TimeoutExpired, OSError):
        return {}

    result: dict[str, int] = {}
    for line in r.stdout.splitlines():
        if ":" not in line:
            continue
        key = line.split(":")[0].strip()
        val = line.split(":", 1)[1].strip()
        try:
            result[key] = int(val)
        except ValueError:
            continue
    return result


def _sysctl_value(key: str) -> int | None:
    """Read a single sysctl key as an integer."""
    try:
        r = subprocess.run(
            ["/usr/sbin/sysctl", "-n", key],
            capture_output=True,
            text=True,
            timeout=2,
        )
        if r.returncode != 0:
            return None
        val = r.stdout.strip()
        if val.startswith("{"):
            nums = [int(x) for x in val.strip("{}").split() if x.lstrip("-").isdigit()]
            return sum(nums) if nums else None
        if ":" in val and not val.replace(":", "").replace(" ", "").replace("-", "").isdigit():
            return None
        return int(val)
    except (ValueError, subprocess.TimeoutExpired, OSError):
        return None


class SysctlSource(EntropySource):
    """Harvest entropy from rapidly-changing sysctl kernel counters.

    On first use, probes all numeric sysctl keys and discovers which ones
    fluctuate.  Subsequent calls sample only the fluctuating keys and
    extract entropy from their deltas.
    """

    name = "sysctl_counters"
    description = "Kernel counter deltas from 50+ fluctuating sysctl keys"
    platform_requirements = ["darwin"]
    entropy_rate_estimate = 5000.0  # very high — many independent counters

    def __init__(self) -> None:
        self._fluctuating_keys: list[str] | None = None

    def is_available(self) -> bool:
        if platform.system() != "Darwin":
            return False
        # Check sysctl works by reading any numeric key
        return _sysctl_value("kern.osrelease") is not None or _sysctl_value("hw.ncpu") is not None

    def discover_fluctuating_keys(self, probe_delay: float = 0.2) -> list[str]:
        """Probe all numeric sysctl keys; return those that change.

        Uses batch ``sysctl -a`` (2 calls total) instead of per-key probing.
        """
        snap1 = _sysctl_snapshot()
        if not snap1:
            return []

        time.sleep(probe_delay)

        snap2 = _sysctl_snapshot()
        changing = [k for k in snap1 if k in snap2 and snap2[k] != snap1[k]]

        self._fluctuating_keys = sorted(changing)
        return self._fluctuating_keys

    @property
    def fluctuating_keys(self) -> list[str]:
        if self._fluctuating_keys is None:
            self.discover_fluctuating_keys()
        return self._fluctuating_keys  # type: ignore[return-value]

    def categorize_keys(self) -> dict[str, list[str]]:
        """Group fluctuating keys by category prefix."""
        cats: dict[str, list[str]] = {}
        for k in self.fluctuating_keys:
            prefix = k.split(".")[0]
            cats.setdefault(prefix, []).append(k)
        return cats

    def collect(self, n_samples: int = 1000) -> np.ndarray:
        """Sample fluctuating sysctl keys and extract delta LSBs.

        Uses batch ``sysctl -a`` snapshots for speed (~0.1s each vs minutes
        when reading keys individually).
        """
        keys = self.fluctuating_keys
        if not keys:
            return np.array([], dtype=np.uint8)

        key_set = set(keys)
        rounds = max(2, n_samples // max(len(keys), 1) + 1)
        prev: dict[str, int] | None = None
        deltas: list[int] = []

        for _ in range(rounds):
            snap = _sysctl_snapshot()
            if prev is not None:
                for k in keys:
                    if k in snap and k in prev:
                        d = snap[k] - prev[k]
                        if d != 0:
                            deltas.append(d)
            prev = {k: v for k, v in snap.items() if k in key_set}
            if len(deltas) >= n_samples:
                break

        if not deltas:
            return np.array([], dtype=np.uint8)

        arr = np.array(deltas[:n_samples], dtype=np.int64)
        return (arr & 0xFF).astype(np.uint8)

    def entropy_quality(self) -> dict:
        data = self.collect(1000)
        q = self._quick_quality(data, self.name)
        q["fluctuating_keys"] = len(self.fluctuating_keys)
        q["categories"] = self.categorize_keys()
        return q

File: crates/openentropy-core/src/sources/sysctl.rs
//! SysctlSource — Batch-reads kernel counters via `sysctl -a`, takes multiple
//! snapshots, finds keys that change between snapshots, extracts deltas of
//! changing values, XORs consecutive deltas, and extracts LSBs.

use std::collections::HashMap;
use std::thread;
use std::time::Duration;

use crate::source::{EntropySource, SourceCategory, SourceInfo};

use super::helpers::{extract_delta_bytes_i64, run_command};

/// Path to the sysctl binary on macOS.
const SYSCTL_PATH: &str = "/usr/sbin/sysctl";

/// Delay between the two sysctl snapshots.
const SNAPSHOT_DELAY: Duration = Duration::from_millis(100);

pub struct SysctlSource {
    info: SourceInfo,
}

impl SysctlSource {
    pub fn new() -> Self {
        Self {
            info: SourceInfo {
                name: "sysctl_deltas",
                description: "Batch-reads ~1600 kernel counters via sysctl -a and extracts deltas from the ~40-60 that change within 200ms",
                physics: "Batch-reads ~1600 kernel counters via sysctl and extracts deltas from \
                    the ~40-60 that change within 200ms. These counters track page faults, context \
                    switches, TCP segments, interrupts \u{2014} each driven by independent processes. \
                    The LSBs of their deltas reflect the unpredictable micro-timing of the entire \
                    operating system\u{2019}s activity.",
                category: SourceCategory::System,
                platform_requirements: &["macos"],
                entropy_rate_estimate: 5000.0,
            },
        }
    }
}

impl Default for SysctlSource {
    fn default() -> Self {
        Self::new()
    }
}

/// Run `sysctl -a` and parse every line that has a numeric value into a HashMap.
///
/// Handles both `key: value` (macOS) and `key = value` (Linux) formats.
fn snapshot_sysctl() -> Option<HashMap<String, i64>> {
    let stdout = run_command(SYSCTL_PATH, &["-a"])?;
    let mut map = HashMap::new();

    for line in stdout.lines() {
        // Try "key: value" first (macOS style), then "key = value" (Linux style)
        let (key, val_str) = if let Some(idx) = line.find(": ") {
            (&line[..idx], line[idx + 2..].trim())
        } else if let Some(idx) = line.find(" = ") {
            (&line[..idx], line[idx + 3..].trim())
        } else {
            continue;
        };

        // Only keep entries whose value is a plain integer
        if let Ok(v) = val_str.parse::<i64>() {
            map.insert(key.to_string(), v);
        }
    }

    Some(map)
}

impl EntropySource for SysctlSource {
    fn info(&self) -> &SourceInfo {
        &self.info
    }

    fn is_available(&self) -> bool {
        std::path::Path::new(SYSCTL_PATH).exists()
    }

    fn collect(&self, n_samples: usize) -> Vec<u8> {
        // Take two snapshots separated by a small delay
        let snap1 = match snapshot_sysctl() {
            Some(s) => s,
            None => return Vec::new(),
        };

        thread::sleep(SNAPSHOT_DELAY);

        let snap2 = match snapshot_sysctl() {
            Some(s) => s,
            None => return Vec::new(),
        };

        // Find keys that changed between the two snapshots and compute deltas
        let mut deltas: Vec<i64> = Vec::new();
        for (key, v2) in &snap2 {
            if let Some(v1) = snap1.get(key) {
                let delta = v2.wrapping_sub(*v1);
                if delta != 0 {
                    deltas.push(delta);
                }
            }
        }

        extract_delta_bytes_i64(&deltas, n_samples)
    }
}
